webapp/src/serial.tsx

```diff
@@ -0,0 +1,452 @@
+            if (this.chartRoot) this.chartRoot.classList.remove("noconsole");
+            if (this.consoleRoot) this.consoleRoot.classList.remove("noconsole");
+        }
+    }
+
+    dropStaleCharts() {
+        let now = Util.now()
+        this.charts.forEach((chart) => {
+            if (now - chart.lastUpdatedTime > this.maxChartTime) {
+                this.chartRoot.removeChild(chart.rootElement)
+                chart.isStale = true
+            }
+        })
+        this.charts = this.charts.filter(c => !c.isStale)
+    }
+
+    pauseRecording() {
+        this.active = false
+        if (this.startPauseButton) this.startPauseButton.setState({ active: this.active });
+        this.charts.forEach(s => s.stop())
+    }
+
+    startRecording() {
+        this.active = true
+        if (this.startPauseButton) this.startPauseButton.setState({ active: this.active });
+        this.charts.forEach(s => s.start())
+    }
+
+    toggleRecording() {
+        pxt.tickEvent("serial.toggleRecording")
+        if (this.active) this.pauseRecording()
+        else this.startRecording()
+    }
+
+    clearNode(e: HTMLElement) {
+        while (e.hasChildNodes()) {
+            e.removeChild(e.firstChild)
+        }
+    }
+
+    clear() {
+        if (this.chartRoot) this.clearNode(this.chartRoot)
+        if (this.clearNode) this.clearNode(this.consoleRoot)
+        if (this.chartRoot) {
+            this.clearNode(this.chartRoot);
+            this.chartRoot.classList.add("noconsole")
+        }
+        if (this.consoleRoot) {
+            this.clearNode(this.consoleRoot);
+            this.consoleRoot.classList.add("noconsole")
+        }
+        this.charts = []
+        this.consoleBuffer = ""
+    }
+
+    entriesToPlaintext() {
+        return this.rawDataBuffer
+    }
+
+    entriesToCSV() {
+        let csv = this.charts.map(chart => `time (s), ${chart.variable} (${chart.source})`).join(', ') + '\r\n';
+        const datas = this.charts.map(chart => chart.line.data);
+        const nl = datas.map(data => data.length).reduce((l, c) => Math.max(l, c));
+        const nc = this.charts.length;
+        for (let i = 0; i < nl; ++i) {
+            csv += datas.map(data => i < data.length ? `${(data[i][0] - data[0][0]) / 1000}, ${data[i][1]}` : ' , ').join(', ');
+            csv += '\r\n';
+        }
+        return csv;
+    }
+
+    showExportDialog() {
+        pxt.tickEvent("serial.showExportDialog")
+        const targetTheme = pxt.appTarget.appTheme
+        let rootUrl = targetTheme.embedUrl
+        if (!rootUrl) {
+            pxt.commands.browserDownloadAsync(this.entriesToPlaintext(), "data.txt", "text/plain")
+            return
+        }
+        if (!/\/$/.test(rootUrl)) rootUrl += '/'
+
+        core.confirmAsync({
+            logos: undefined,
+            header: lf("Export data"),
+            hideAgree: true,
+            disagreeLbl: lf("Close"),
+            onLoaded: (_) => {
```

Comment:
This file can have more comments. Do be blocked on this. Add when you have time.